core/lib/Hooks.re

```diff
@@ -0,0 +1,266 @@
+type hook('a) = ..;
```

Comment:
Just curious, what's the advantage of the "open variant" approach? as opposed to define `hook('a)` after the modules?

Reply:
I don't know... yet. I thought it might be an interesting idea to implement hooks as an open variant to allow users creating custom types of hooks in the userspace. Like what if we exposed enough internals (but not too much!) to implement `Hooks.suspense` in the user space? I think it could be very interesting.